pkg/pillar/types/global.go

```diff
@@ -103,312 +109,439 @@ func (gs *GlobalStatus) UpdateItemValuesFromGlobalConfig(gc GlobalConfig) {
-	}
-	if newgc.NetworkTestTimeout == 0 {
-		newgc.NetworkTestTimeout = GlobalConfigDefaults.NetworkTestTimeout
+func (specMap *ConfigItemSpecMap) AddBoolItem(key GlobalSettingKey, agentSettingKey AgentSettingKey, defaultBool bool) {
+	configItem := ConfigItemSpec{
+		ItemType:    ConfigItemTypeBool,
+		Key:         string(key),
+		BoolDefault: defaultBool,
 	}
-	if newgc.NetworkSendTimeout == 0 {
-		newgc.NetworkSendTimeout = GlobalConfigDefaults.NetworkSendTimeout
+	if agentSettingKey != "" {
+		configItem.Key = string(agentSettingKey)
+		specMap.AgentSettings[agentSettingKey] = configItem
 	}
-	if newgc.StaleConfigTime == 0 {
-		newgc.StaleConfigTime = GlobalConfigDefaults.StaleConfigTime
+	if key != "" {
+		specMap.GlobalSettings[key] = configItem
 	}
-	if newgc.DownloadGCTime == 0 {
-		newgc.DownloadGCTime = GlobalConfigDefaults.DownloadGCTime
-	}
-	if newgc.VdiskGCTime == 0 {
-		newgc.VdiskGCTime = GlobalConfigDefaults.VdiskGCTime
+	log.Debugf("Added bool item %s", key)
+}
+
+func (specMap *ConfigItemSpecMap) AddStringItem(key GlobalSettingKey, agentSettingKey AgentSettingKey, defaultString string) {
+	configItem := ConfigItemSpec{
+		ItemType:      ConfigItemTypeString,
+		Key:           string(key),
+		StringDefault: defaultString,
 	}
-	if newgc.DownloadRetryTime == 0 {
-		newgc.DownloadRetryTime = GlobalConfigDefaults.DownloadRetryTime
+	if agentSettingKey != "" {
+		configItem.Key = string(agentSettingKey)
+		specMap.AgentSettings[agentSettingKey] = configItem
 	}
-	if newgc.DomainBootRetryTime == 0 {
-		newgc.DomainBootRetryTime = GlobalConfigDefaults.DomainBootRetryTime
+	if key != "" {
+		specMap.GlobalSettings[key] = configItem
 	}
-	if newgc.DefaultLogLevel == "" {
-		newgc.DefaultLogLevel = GlobalConfigDefaults.DefaultLogLevel
+	log.Debugf("Added string item %s", key)
+}
+
+func (specMap *ConfigItemSpecMap) AddTriStateItem(key GlobalSettingKey, agentSettingKey AgentSettingKey, defaultTriState TriState) {
+	configItem := ConfigItemSpec{
+		Key:             string(key),
+		ItemType:        ConfigItemTypeTristate,
+		TriStateDefault: defaultTriState,
 	}
-	if newgc.DefaultRemoteLogLevel == "" {
-		newgc.DefaultRemoteLogLevel = GlobalConfigDefaults.DefaultRemoteLogLevel
+	if agentSettingKey != "" {
+		configItem.Key = string(agentSettingKey)
+		specMap.AgentSettings[agentSettingKey] = configItem
 	}
-	if newgc.AllowNonFreeAppImages == TS_NONE {
-		newgc.AllowNonFreeAppImages = GlobalConfigDefaults.AllowNonFreeAppImages
+	if key != "" {
+		specMap.GlobalSettings[key] = configItem
 	}
-	if newgc.AllowNonFreeBaseImages == TS_NONE {
-		newgc.AllowNonFreeBaseImages = GlobalConfigDefaults.AllowNonFreeBaseImages
+	log.Debugf("Added tristate item %s", key)
+}
+
+func (specMap *ConfigItemSpecMap) ParseItem(configMap *ConfigItemValueMap, item config.ConfigItem) error {
+	agentSetting := false
+	var configSpec ConfigItemSpec
+	globalSetting := false
+	stringKey := item.Key
```

Comment:
why stringKey? just key or itemKey would be better? You anyways know key is string..